Sources/Subs-Members.php

```diff
@@ -1376,18 +1376,7 @@ function populateDuplicateMembers(&$members)
 // Generate a random validation code.
 function generateValidationCode()
 {
-	global $smcFunc, $modSettings;
-
-	$request = $smcFunc['db_query']('get_random_number', '
-		SELECT RAND()',
-		array(
-		)
-	);
-
-	list ($dbRand) = $smcFunc['db_fetch_row']($request);
-	$smcFunc['db_free_result']($request);
-
-	return substr(preg_replace('/\W/', '', sha1(microtime() . mt_rand() . $dbRand . $modSettings['rand_seed'])), 0, 10);
+	return bin2hex(random_bytes(5));
```

Comment:
Perhaps use `base64_encode` rather than `bin2hex`, as that provides a little more entropy by using all 26 letters in both upper- and lower-case. It then still requires the use of `substr` to pull out the first 10 characters, but I don't think that's a big deal.

Reply:
Several reasons why I chose not to do that:

The original implementation was lower-case only, so I speculate that keeping it that way is less likely to cause problems in the future. This is SMF, at some point they'll introduce some arbitrary function that relies on this entirely too much. 

I'm also not convinced that this provides more entropy - you're proposing a different representation of the same random number, except now you've introduced the possibility of collisions. To my mind, the correct source of entropy is `random_bytes` - if we want more entropy, we ask for more bytes. (But I'd also prefer not to do that without good reason, since I'd like to keep this implementation compatible with hypothetical future bullshit). 

Finally, there are use (exceptionally rare) cases where the user might have to type this code in by hand. In these cases, keeping the representation lower-case might make life slightly easier.

Reply:
My implication was to get more bytes from `random_bytes`, since you need more bytes to produce a base64-encoded string of the same length.

I'm sold by your other points, though. It might still be nice to have the full 26 letters in lower-case, but that also introduces potential ambiguity (e.g., between the letter l and the numeral 1) when typing the code.